init.php

```diff
@@ -1,121 +1,173 @@
-				$hidden_items[ $item->ID ] = '1';
-			}
-		}
-
-		return $items;
-	}
-
-	/**
-	 * Remove the _menu_item_visibility meta when the menu item is removed
-	 *
-	 * @since 0.2.2
-	 */
-	public static function remove_visibility_meta( $post_id ) {
-		if ( is_nav_menu_item( $post_id ) ) {
-			delete_post_meta( $post_id, '_menu_item_visibility' );
-		}
-	}
-
-	/**
-	 * Handle errors in eval-ed Logic field
-	 *
-	 * @since 0.4
-	 */
-	public static function error_handler( $errno, $errstr, $errFile, $errLine, $errContext ) {
-		if ( current_user_can( 'manage_options' ) ) {
-			if ( ! empty( $errContext['item']->title ) ) {
-				echo sprintf( __( 'Error in "%s" menu item Visibility: ', 'menu-items-visibility-control' ), $errContext['item']->title );
-			}
-
-			echo $errstr;
-		}
-
-		/* Don't execute PHP internal error handler */
-		return true;
-	}
+    /**
+     * Constructor.
+     */
+    public function __construct() {
+
+        if ( ! is_admin() ) {
+            add_filter( 'wp_get_nav_menu_items', [ __CLASS__, 'visibility_check' ], 10, 3 );
+            return;
+        }
+
+        add_action( 'wp_nav_menu_item_custom_fields', [ __CLASS__, 'wp_nav_menu_item_custom_fields' ] );
+        add_action( 'wp_update_nav_menu_item', [ __CLASS__, 'update_option' ], 10, 3 );
+        add_action( 'delete_post', [ __CLASS__, 'remove_visibility_meta' ], 1, 3 );
+
+    }
+
+    /**
+     * Display condition field in menu admin.
+     *
+     * @param int $item_id
+     *
+     * @return void
+     *
+     * @since 0.3.8
+     */
+    public static function wp_nav_menu_item_custom_fields( int $item_id ) : void {
+
+        $value = get_post_meta( $item_id, '_menu_item_visibility', true );
+
+        ?>
+        <p class="field-visibility description description-wide">
+            <label for="edit-menu-item-visibility-<?php echo $item_id; ?>">
+                <?php printf( __( 'Visibility logic (<a href="%s">?</a>)', 'menu-items-visibility-control' ), 'https://codex.wordpress.org/Conditional_Tags' ); ?><br>
+                <input type="text" class="widefat code" id="edit-menu-item-visibility-<?php echo $item_id; ?>" name="menu-item-visibility[<?php echo esc_attr( $item_id ); ?>]" value="<?php echo esc_attr( $value ); ?>" />
+            </label>
+        </p>
+        <?php
+
+    }
+
+    /**
+     * Update menu item visibility post meta.
+     *
+     * @param int $menu_id
+     * @param int $menu_item_db_id
+     * @param array $args
+     *
+     * @return void
+     *
+     * @since 0.3.8
+     */
+    public static function update_option( int $menu_id, int $menu_item_db_id, array $args ) : void {
+
+        $menu_item_visibility = filter_input( INPUT_POST, 'menu-item-visibility' );
+
+        if ( isset( $menu_item_visibility ) ) {
+
+            $meta_value     = get_post_meta( $menu_item_db_id, '_menu_item_visibility', true );
+            $new_meta_value = stripcslashes( $_POST['menu-item-visibility'][ $menu_item_db_id ] );
```

Comment:
How come direct `$_POST` access here? Didn't we just get this value into `$menu_item_visibility`?

Reply:
filter_input() didnt return array properly as $_POST did. now per longer research i found it needed the fourth parameter "options" added, `FILTER_REQUIRE_ARRAY`.